# **01 - Data Collection**

## Objectives

* Install the Kaggle API to enable dataset retrieval.
* Authenticate using the Kaggle API token stored in the `kaggle.json` file.
* Download the raw dataset from Kaggle.
* Inspect the dataset to ensure it is loaded correctly.
* Save the dataset in the `inputs/datasets/raw/` folder for further processing.

## Inputs

* Kaggle dataset: [Housing Prices Data](
)
* Kaggle API token stored in the `kaggle.json` file.

## Outputs

* Raw dataset saved as `inputs/datasets/raw/housing_prices.csv`.

## Additional Comments

* Ensure that the Kaggle API token (`kaggle.json`) is correctly configured and stored in the project root directory.
* The working directory will be adjusted to the project root to ensure consistent file paths.

---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/workspaces/milestone-project-heritage-housing-issues/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/workspaces/milestone-project-heritage-housing-issues'

# Section 1: Data Retrieval

To retrieve the raw project data, we first need to install the Kaggle API, which allows us to download datasets directly from Kaggle. This requires the installation of the Kaggle Python package. 

In [4]:
%pip install kaggle


[notice] A new release of pip is available: 24.3.1 -> 25.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


To configure the Kaggle API for data set retrieval, we first need to set the enviroment variable `KAGGLE_CONFIG_DIR` to the current directory where the `kaggle.json` file is stored. This file contains the necessary credentials for authentification. 
We use the `chmod 600` to restrict the access to the `kaggle.json` file, ensuring that it is only readable and writeble by the owner.

In [5]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = os.getcwd()
! chmod 600 kaggle.json

Next, we download the zip file containing the dataset.

In [6]:
KaggleDatasetPath = "codeinstitute/housing-prices-data"
DestinationFolder = "inputs/datasets/raw"   
! kaggle datasets download -d {KaggleDatasetPath} -p {DestinationFolder}

Dataset URL: https://www.kaggle.com/datasets/codeinstitute/housing-prices-data
License(s): unknown
  0%|                                               | 0.00/49.6k [00:00<?, ?B/s]
100%|███████████████████████████████████████| 49.6k/49.6k [00:00<00:00, 113MB/s]


Unzip dataset, delete zip-file and remove kaggle.json file.

In [16]:
! unzip {DestinationFolder}/*.zip -d {DestinationFolder} \
  && rm {DestinationFolder}/*.zip \
  && rm kaggle.json

unzip:  cannot find or open inputs/datasets/raw/*.zip, inputs/datasets/raw/*.zip.zip or inputs/datasets/raw/*.zip.ZIP.

No zipfiles found.


Moving the CSV files to `/inputs/datasets/raw/` and removing unnecessary folders simplifies the project structure, makes the code cleaner, and follows best practices for organizing raw data.

In [17]:
!mv inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/*.csv inputs/datasets/raw/ \
  && rmdir inputs/datasets/raw/house-price-20211124T154130Z-001/house-price \
  && rmdir inputs/datasets/raw/house-price-20211124T154130Z-001

---

# Section 2: Load and Inspect Kaggle Data

* Import the Pandas library to work with tabular data.
* Load the datasets (`house_prices_records.csv`and `inherited_houses.csv`) into two separate Pandas DataFrames.
* Display the first few rows of `df_prices`.

In [36]:
import pandas as pd

df_prices = pd.read_csv("inputs/datasets/raw/house_prices_records.csv")
df_inherited = pd.read_csv("inputs/datasets/raw/inherited_houses.csv")

df_prices.head()

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       856     854.0           3.0           No         706          GLQ   
1      1262       0.0           3.0           Gd         978          ALQ   
2       920     866.0           3.0           Mn         486          GLQ   
3       961       NaN           NaN           No         216          ALQ   
4      1145       NaN           4.0           Av         655          GLQ   

   BsmtUnfSF  EnclosedPorch  GarageArea GarageFinish  ...  LotFrontage  \
0        150            0.0         548          RFn  ...         65.0   
1        284            NaN         460          RFn  ...         80.0   
2        434            0.0         608          RFn  ...         68.0   
3        540            NaN         642          Unf  ...         60.0   
4        490            0.0         836          RFn  ...         84.0   

   MasVnrArea OpenPorchSF  OverallCond  OverallQual  TotalBsmtSF  WoodDeckSF  \
0       196.0          61            5            7          856         0.0   
1         0.0           0            8            6         1262         NaN   
2       162.0          42            5            7          920         NaN   
3         0.0          35            5            7          756         NaN   
4       350.0          84            5            8         1145         NaN   

   YearBuilt  YearRemodAdd  SalePrice  
0       2003          2003     208500  
1       1976          1976     181500  
2       2001          2002     223500  
3       1915          1970     140000  
4       2000          2000     250000  

[5 rows x 24 columns]

Inspecting DataFrame Structure.
We use `.info()` to inspect the structure of the DataFrame, including column names, data types, and the number of non-null values. This helps us identify missing values and understand the overall structure of the dataset.

In [43]:
print("Dataset info:")
df_prices.info()

Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype   
---  ------         --------------  -----   
 0   1stFlrSF       1460 non-null   int64   
 1   2ndFlrSF       1374 non-null   float64 
 2   BedroomAbvGr   1361 non-null   float64 
 3   BsmtExposure   1422 non-null   category
 4   BsmtFinSF1     1460 non-null   int64   
 5   BsmtFinType1   1315 non-null   category
 6   BsmtUnfSF      1460 non-null   int64   
 7   EnclosedPorch  136 non-null    float64 
 8   GarageArea     1460 non-null   int64   
 9   GarageFinish   1225 non-null   category
 10  GarageYrBlt    1379 non-null   float64 
 11  GrLivArea      1460 non-null   int64   
 12  KitchenQual    1460 non-null   category
 13  LotArea        1460 non-null   int64   
 14  LotFrontage    1201 non-null   float64 
 15  MasVnrArea     1452 non-null   float64 
 16  OpenPorchSF    1460 non-null   int64   
 17  OverallCond    1460

Convert categorical columns to `category`.

In [45]:
categorical_columns = ['BsmtExposure', 'BsmtFinType1', 'GarageFinish', 'KitchenQual']
for col in categorical_columns:
    df_prices[col] = df_prices[col].astype('category')

Identify the number of missing values in each column of `df_prices`. To count the number of missing values in each column and plan how we will handle them.

In [40]:
df_prices.isnull().sum()

1stFlrSF            0
2ndFlrSF           86
BedroomAbvGr       99
BsmtExposure       38
BsmtFinSF1          0
BsmtFinType1      145
BsmtUnfSF           0
EnclosedPorch    1324
GarageArea          0
GarageFinish      235
GarageYrBlt        81
GrLivArea           0
KitchenQual         0
LotArea             0
LotFrontage       259
MasVnrArea          8
OpenPorchSF         0
OverallCond         0
OverallQual         0
TotalBsmtSF         0
WoodDeckSF       1305
YearBuilt           0
YearRemodAdd        0
SalePrice           0
dtype: int64

Statistical Summary of Numeric Columns.
We use `.describe()` to generate a statistical summary of the numeric columns in the dataset. This helps us understand the distribution of data, identify outliers, and plan for data cleaning.

In [41]:
print("Statistical Summary of df_prices:")
print(df_prices.describe())

Statistical Summary of df_prices:
          1stFlrSF     2ndFlrSF  BedroomAbvGr   BsmtFinSF1    BsmtUnfSF  \
count  1460.000000  1374.000000   1361.000000  1460.000000  1460.000000   
mean   1162.626712   348.524017      2.869214   443.639726   567.240411   
std     386.587738   438.865586      0.820115   456.098091   441.866955   
min     334.000000     0.000000      0.000000     0.000000     0.000000   
25%     882.000000     0.000000      2.000000     0.000000   223.000000   
50%    1087.000000     0.000000      3.000000   383.500000   477.500000   
75%    1391.250000   728.000000      3.000000   712.250000   808.000000   
max    4692.000000  2065.000000      8.000000  5644.000000  2336.000000   

       EnclosedPorch   GarageArea  GarageYrBlt    GrLivArea        LotArea  \
count     136.000000  1460.000000  1379.000000  1460.000000    1460.000000   
mean       25.330882   472.980137  1978.506164  1515.463699   10516.828082   
std        66.684115   213.804841    24.689725   525.480

---

## Conclusions and next step
### Concolusions:
**From `.info()`:**
* Columns with missing value: `2ndFlrSF`, `BedroomAbvGr`, `BsmtExposure`, `BsmtFinType1`, `EnclosedPorch`, `GarageFinish`, `GarageYrBlt`, `LotFrontage`, `MasVnrArea`, `WoodDeckSF`.
* Categorical columns (`object`) were converted to `category` for optimization. 
* The dataset has 1460 rows and 24 columns.

**From `.isnull().sum()`:**
* `LotFrontage` has 259 missing values (17.7% of the data).
* `GarageFinish` has 235 missing values (16.1% of the data).
* `EnclosedPorch`and `WoodDeckSF`have a high percentage of missing values. 

**From `.describe()`:**
* `SalePrice` has a maximum value of 755000, which might indicate outliners.
* `LotFrontage` has a wide range of values, with a mean of 70 and a max of 313.
* `GarageYrBlt` has missing values but represents years, so it might need special handeling. 

#### Next Steps: Data Cleaning

